reactor-core/src/main/java/reactor/core/publisher/HookableScheduledExecutorService.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2019-Present Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.core.publisher;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+class HookableScheduledExecutorService implements ScheduledExecutorService {
+
+	final ScheduledExecutorService delegate;
+
+	protected HookableScheduledExecutorService(ScheduledExecutorService delegate) {
+		this.delegate = delegate;
+	}
+
+	protected Runnable wrap(Runnable runnable) {
+		return Hooks.onScheduleHook.apply(runnable);
+	}
+
+	protected <V> Runnable wrap(Callable<V> callable) {
+		// We assume that every Callable in Reactor implements Runnable
+		return Hooks.onScheduleHook.apply((Runnable) callable);
+	}
+
+	@Override
+	public ScheduledFuture<?> schedule(Runnable command,
+			long delay,
+			TimeUnit unit) {
+		return delegate.schedule(wrap(command), delay, unit);
+	}
+
+	@Override
+	public <V> ScheduledFuture<V> schedule(Callable<V> callable,
+			long delay,
+			TimeUnit unit) {
+		return (ScheduledFuture<V>) delegate.schedule(wrap(callable), delay, unit); //FIXME
```

Comment:
@bsideup @smaldini I can see why the new interface was proposed to map both `Runnable` and `Callable`. Maybe instead of hard casting we should throw UOE, saying that reactor internals are not supposed to call `ScheduledExecutorService` Callable-based methods?

And probably hunt down every place where we have access to an executor to introduce a comment that such executors' Callable-based methods are off limits?

Reply:
Since we leak the `ScheduledExecutorService` to our public API, these UOE make me feel uneasy to be honest. Did you come to some agreement with @smaldini about this?